File: Stack-SlidingWindow-Queue/2552-maximum-sum-of-distinct-subarrays-with-length-k/solution.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        unordered_map <long long,long long> mp;
        long long mx = 0, n=nums.size(),running_sum=0, flag =0;
        for(int i=0;i<k;i++){
            mp[nums[i]]++;
            if(mp[nums[i]]>1) flag +=1;
            running_sum+=nums[i];
        }
        if(flag==0) mx= max(running_sum,mx);
        for(int i = k;i<n;i++){
            if(--mp[nums[i-k]]>0) flag --;
            if(++mp[nums[i]]>1) flag++;
            running_sum+= nums[i];
            running_sum-=nums[i-k];
            if(flag==0) mx=max(running_sum,mx);
        }
        return mx;
    }
};

File: Stack-SlidingWindow-Queue/2552-maximum-sum-of-distinct-subarrays-with-length-k/solution.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        dict1=defaultdict(int)
        mx=0
        rs=0
        n= len(nums)
        flag =0
        for i in range(k):
            if nums[i] in dict1:
                flag+=1
                dict1[nums[i]]+=1
            else:
                dict1[nums[i]]=1
            rs+= nums[i]
        if flag==0:
            mx=max(rs,mx)
        for i in range(k,n,1):
            dict1[nums[i-k]]-=1
            if dict1[nums[i-k]]>0:
                flag -=1
            if nums[i] in dict1:
                dict1[nums[i]]+=1
            else:
                dict1[nums[i]]=1
            if dict1[nums[i]] >1:
                flag +=1
            rs+=(nums[i]-nums[i-k])
            if flag==0:
                mx = max(mx,rs)
        return mx

